### PR TITLE
LOG-5519: Refactor outputs to support templating

### PR DESCRIPTION
**Pull request by Clee2691**
### Description
This PR refactors outputs with supported tenancy configuration to allow a limited version of vector's [template syntax](https://vector.dev/docs/reference/configuration/template-syntax/).

This also removes AWS Cloudwatch's `groupPrefix` & `groupBy` and condenses them into `groupName`.

Supported outputs are:
1. Cloudwatch `groupName`
2. Elasticsearch `index`
3. GoogleCloudLogging `logId`
4. Kafka `topic`
5. Loki `tenantKey`
6. Splunk `index`

A static tenant must match the pattern consisting of only letters, numbers, underscores (`_`), dots (`.`), dashes (`-`), and slashes(`/`).
For dynamic values, a path expression must only include letters, numbers, underscores (`_`), and dots (`.`). A space can be added before and after the path expression like so: `{{ .log_type }}`.

Valid Tenant Examples:
1. `baz-{{.foo.bar}}`
2. `foo`
3. `{{.foo}}`
4. `foo-bar_{{.baz}}`
5. `foo-{{.bar.baz}}-foo`
6. `{{.foo}}-{{.bar}}_{{.foo.bar.baz}}`
7. `{{.foo}}{{.bar}}`
8. `{{ .foo }}/{{ .bar }}`

/cc @cahartma @vparfonov 
/assign @jcantrill 

### Links
- JIRA: https://issues.redhat.com/browse/LOG-5519

### CODE REVIEW

File: test/framework/functional/output_cloudwatch.go [hunk 1]
```diff
@@ -141,7 +142,7 @@ func (f *CollectorFunctionalFramework) GetLogGroupByType(inputName string) ([]st
 	found := false
 	for _, l := range logGroupsOutput.LogGroups {
 		// Filter by type and get all
-		if *l.LogGroupName == "group-prefix."+inputName {
+		if *l.LogGroupName == "group-prefix-"+inputName {
```
Comment (jcantrill) [inline test/framework/functional/output_cloudwatch.go:145]: why this chane?

**Comment by jcantrill** [comment]
@Clee2691 lets make sure we identify changes from before and are prepared to properly document them

File: api/observability/v1/output_types.go [hunk 2]
```diff
@@ -447,6 +424,10 @@ type GoogleCloudLogging struct {
 	ID GoogleGloudLoggingID `json:"id,omitempty"`
 
 	// LogID is the log ID to which to publish logs. This identifies log stream.
+	//
+	// +kubebuilder:validation:Pattern:=`^([a-zA-Z0-9]+([-_])?)*(\{\{\.[a-zA-Z0-9_.]+?\}\}([-_])?)*([a-zA-Z0-9]+([-_])?)*$`
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:="{{.log_type}}"
```
[the remark this answers is not in the document]
Comment (Clee2691) [reply api/observability/v1/output_types.go:430]: Well according to vector's docs, the [log_id](https://vector.dev/docs/reference/configuration/sinks/gcp_stackdriver_logs/#log_id) is required. I viewed this similar to Cloudwatch's `group_name` which also describes a log stream and is also required and were one of:
```
1. .kubernetes.namespace_name
2. .kubernetes.namespace_id
3. .log_type
```

File: internal/generator/vector/output/cloudwatch/cloudwatch.go [hunk 3]
```diff
@@ -128,31 +129,19 @@ func endpointConfig(cw *obs.Cloudwatch) Element {
...
 	vrl := strings.TrimSpace(`
-.group_name = "default"
 .stream_name = "default"
 
 if (.file != null) {
  .file = "kubernetes" + replace!(.file, "/", ".")
  .stream_name = del(.file)
 }
 
-if ( .log_type == "application" ) {
```
[the remark this answers is not in the document]
Comment (Clee2691) [reply internal/generator/vector/output/cloudwatch/cloudwatch.go:147]: Noted

File: internal/generator/vector/output/cloudwatch/cloudwatch.go [hunk 4]
```diff
@@ -128,31 +129,19 @@ func endpointConfig(cw *obs.Cloudwatch) Element {
...
-	appGroupName := fmt.Sprintf("%q + %s", logGroupPrefix, logGroupNameField)
-	if logGroupPrefix == "" {
-		appGroupName = logGroupNameField
-	}
-	auditGroupName := fmt.Sprintf("%s%s", logGroupPrefix, "audit")
-	infraGroupName := fmt.Sprintf("%s%s", logGroupPrefix, "infrastructure")
+func NormalizeStreamName(componentID string, inputs []string) Element {
 	vrl := strings.TrimSpace(`
-.group_name = "default"
 .stream_name = "default"
```
[the remark this answers is not in the document]
Comment (Clee2691) [reply internal/generator/vector/output/cloudwatch/cloudwatch.go:134]: I'm not sure what you mean by this. The `group_name` identifies the log stream and is what is now `groupName` in the API that combined `groupBy` and `groupPrefix`.
If you mean the `stream_name` I can see us templating this but wasn't something we exposed in the API originally. `stream_name` is separate from the `group_name`.

File: test/framework/functional/output_cloudwatch.go (hunk 1 above)
Comment (Clee2691) [reply test/framework/functional/output_cloudwatch.go:145]: I guess the bigger question here is what characters are allowed/not allowed for this as well as indices for other outputs.

**Pull request merged**